race-buffer/src/lib.rs

```diff
@@ -0,0 +1,303 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use core::cmp::PartialEq;
+use core::marker::Copy;
+
+/// Entry that can be stored in a RaceBuffer
+pub trait Entry: Copy + PartialEq {
+    /// Return true if entry is the first in a double entry
+    fn is_prefix(&self) -> bool;
+}
+
+/// Returns the corresponding index in backing storage of given cursor index
+#[inline]
+fn get_cursor_index(storage_cap: usize, cursor: usize, use_base_2_indexing: bool) -> usize {
+    if use_base_2_indexing {
+        // Index is lowest n bits of cursor where storage_cap is 2^n
+        cursor & (storage_cap - 1)
+    } else {
+        cursor % storage_cap
+    }
+}
+
+pub mod writer;
+
+#[cfg(feature = "std")]
+pub mod reader;
+
+#[cfg(feature = "std")]
+#[cfg(test)]
+pub mod tests {
```

Comment:
Could I request some tests that capture the exact expected behavior when we write more than the buffer can store? In that same vein, tests that show what happens when we attempt to read when the buffer is empty? Not-quite-equivalently, what we expect to happen when we are both reading and writing but doing one of those more often than the other.  IIUC, said tests wouldn't necessarily have to be multithreaded to be useful codifications.

Reply:
cc @NicholasSpringer ^^

Reply:
@ZackPierce Just so you know, there is another branch off of `reporty-unity` called `race-buffer-modifications` with quite a few changes to the race buffer. It has some more tests and other modifications, but the things you have suggested so far still need to be addressed.